platforms/nuttx/src/px4/microchip/mpfs/px4io_serial/px4io_serial.cpp

```diff
@@ -0,0 +1,379 @@
+
+	putreg32(0, PX4IO_SERIAL_BASE + MPFS_UART_TTG_OFFSET);
+
+	/* 8 bits, no parity 1 stop */
+
+	uint32_t lcr = UART_LCR_DLS_8BITS;
+
+	/* Set bits, enter dlab */
+
+	putreg32(lcr | UART_LCR_DLAB, PX4IO_SERIAL_BASE + MPFS_UART_LCR_OFFSET);
+
+	/*  Baud_value = PCLK_Frequency / (baud_rate * 16)
+	 *
+	 *  Calculate in fixed point 28.6 format (fraction is presented by 6 bits in PF)
+	 */
+
+	uint32_t baud28_6 = ((uint32_t)MPFS_FPGA_PERIPHERAL_CLK << 2) / PX4IO_SERIAL_BITRATE;
+	uint16_t baud = baud28_6 >> 6;
+	uint8_t fraction = baud28_6 & 0x3F;
+
+	putreg32(baud >> 8, PX4IO_SERIAL_BASE + MPFS_UART_DLH_OFFSET);
+	putreg32(baud & 0xff, PX4IO_SERIAL_BASE + MPFS_UART_DLL_OFFSET);
+
+	if (baud > 1) {
+		/* Enable Fractional baud rate */
+
+		uint8_t mm0 = getreg32(PX4IO_SERIAL_BASE + MPFS_UART_MM0_OFFSET);
+		mm0 |= UART_MM0_EFBR;
+		putreg32(mm0, PX4IO_SERIAL_BASE + MPFS_UART_MM0_OFFSET);
+		putreg32(fraction, PX4IO_SERIAL_BASE + MPFS_UART_DFR_OFFSET);
+	}
+
+	/* Clear dlab */
+
+	putreg32(lcr, PX4IO_SERIAL_BASE + MPFS_UART_LCR_OFFSET);
+
+	/* Configure the FIFO's */
+
+	putreg32((UART_FCR_XFIFOR | UART_FCR_RFIFOR |
+		  UART_FCR_FIFOE), PX4IO_SERIAL_BASE + MPFS_UART_FCR_OFFSET);
+
+	/* Attach serial interrupt handler */
+
+	irq_attach(PX4IO_SERIAL_VECTOR, _interrupt, this);
+	up_enable_irq(PX4IO_SERIAL_VECTOR);
+
+	/* Create semaphores */
+
+	px4_sem_init(&_completion_semaphore, 0, 0);
+
+	/* _completion_semaphore use case is a signal */
+
+	px4_sem_setprotocol(&_completion_semaphore, SEM_PRIO_NONE);
+
+	return 0;
+}
+
+int
+ArchPX4IOSerial::init()
+{
+	/* initialize base implementation */
+	int r = PX4IO_serial::init((IOPacket *)&_io_buffer_storage[0]);
+
+	if (r != 0) {
+		return r;
+	}
+
+	r = init_uart();
+
+	return r;
+}
+
+int
+ArchPX4IOSerial::ioctl(unsigned operation, unsigned &arg)
+{
+	int ret = 0;
+
+	switch (operation) {
+
+	case 1:		/* XXX magic number - test operation */
+		switch (arg) {
+		case 0:
+			syslog(LOG_INFO, "test 0\n");
+
+			while ((getreg32(PX4IO_SERIAL_BASE + MPFS_UART_LSR_OFFSET)
+				& UART_LSR_THRE) == 0);
+
+			putreg32(0x55, PX4IO_SERIAL_BASE + MPFS_UART_THR_OFFSET);
+
+			break;
+
+		case 1: {
+				unsigned fails = 0;
+
+				for (unsigned count = 0;; count++) {
+					uint16_t value = count & 0xffff;
+
+					if (write((PX4IO_PAGE_TEST << 8) | PX4IO_P_TEST_LED, &value, 1) != 0) {
+						fails++;
+					}
+
+					if (count >= 5000) {
+						syslog(LOG_INFO, "==== test 1 : %u failures ====\n", fails);
+						perf_print_counter(_pc_txns);
+						perf_print_counter(_pc_retries);
+						perf_print_counter(_pc_timeouts);
+						perf_print_counter(_pc_crcerrs);
+						perf_print_counter(_pc_protoerrs);
+						perf_print_counter(_pc_uerrs);
+						count = 0;
+					}
+				}
+			}
+			break;
+
+		case 2:
+			syslog(LOG_INFO, "test 2\n");
+			break;
+
+		default:
+			break;
+		}
+
+		break;
+
+	default:
+		ret = -1;
+		break;
+	}
+
+	return ret;
+}
+
+int
+ArchPX4IOSerial::_bus_exchange(IOPacket *_packet)
+{
+
```

Comment:
If you have interest you could remove these extra line feeds. Makes no difference to me. ODC might trigger for some.